ci: integrate kube-api-linter into golangci-lint workflow

moko-poi opened the pull request:
Integrate kube-api-linter into the golangci-lint workflow to enforce Kubernetes API conventions and catch common mistakes in Kubernetes API types.

This PR addresses #675 by:
- Building golangci-lint with the kube-api-linter plugin using `golangci-lint custom`
- Updating CI workflow to use the custom-built linter
- Adding configuration file for custom plugin management

## Changes

### 1. Custom golangci-lint Configuration
- **Added** `.custom-gcl.yml`: Configuration file specifying kube-api-linter plugin (v0.0.0-20251112164541-d94382a24f06) to be built into golangci-lint

### 2. Makefile Updates
- **Modified** `golangci-lint` target (Makefile:214-221): Changed from simple `go install` to custom build process
  - First installs standard golangci-lint v2.4.0
  - Then builds custom binary with plugins using `golangci-lint custom -o $(GOLANGCI_LINT) .custom-gcl.yml`
  - Ensures kube-api-linter is available in local development

### 3. CI/CD Integration
- **Modified** `.github/workflows/golangci-lint.yaml`: Replaced `golangci/golangci-lint-action` with Makefile-based approach
  - Added "Build custom golangci-lint" step to build the linter with plugins
  - Changed "lint" step to use `make lint` for consistency between local and CI environments

## Benefits

- **Standards Compliance**: Enforces [Kubernetes API Conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md)
automatically
- **Consistency**: Same linter configuration used in both local development and CI
- **Maintainability**: Centralized plugin configuration in `.custom-gcl.yml` for easy updates

## Testing

- Local testing: Run `make golangci-lint && make lint` to verify custom linter builds and runs correctly
- CI will automatically build and use the custom linter on all PRs and pushes

## References
- [kubernetes-sigs/kube-api-linter](https://github.com/kubernetes-sigs/kube-api-linter)
- [golangci-lint custom plugins documentation](https://golangci-lint.run/usage/configuration/#custom-plugins)

moko-poi committed:
ci: integrate kube-api-linter into golangci-lint workflow

yorugac commented:
@moko-poi, hello! Would you like to clean up / continue with this PR? As mentioned in the issue, it had [failing CI](https://github.com/grafana/k6-operator/issues/675#issuecomment-3639030426) back then and by now, it also needs a rebase :sweat_smile: 
PS just in case, a forewarning: all the commits need to be signed now! New policy; can't be helped.

moko-poi committed:
ci: enable kubeapilinter with staged rule rollout

Add .golangci.yml to actually enable the kube-api-linter plugin that the
custom golangci-lint build provides. Without this config the plugin was
built but never run.

- Scope kubeapilinter to api/ type definitions only (path-except), so it
  does not flag ordinary internal Go structs under pkg/.
- Enable the linter with a staged rollout: rules with many pre-existing
  violations (commentstart, optionalorrequired, optionalfields,
  arrayofstruct, ssatags) are disabled for now and will be enabled
  incrementally in follow-ups.
- Fix the statusoptional and conditions violations that are now enforced:
  mark TestRunStatus/PrivateLoadZoneStatus fields optional and add the
  standard listType/patchStrategy markers to the Conditions fields.
- Regenerate CRDs accordingly (x-kubernetes-list-type: map on conditions).

moko-poi commented:
Hi @yorugac, thanks for the nudge and for the heads-up about the signing policy. I've gone ahead and cleaned this up.

I rebased the branch onto the latest main and resolved the conflicts, keeping golangci-lint at the current v2.12.1 used on main. All commits are now signed as well.

While doing this I found the real reason the lint job was failing: the previous version only built golangci-lint with the kube-api-linter plugin but never actually enabled it, since there was no .golangci.yml to register the module linter. Once I enabled it properly, the linter reported 118 violations across the existing v1alpha1 API types, which is what was breaking CI.

To avoid turning this into one huge change, I went with a staged rollout. I added a .golangci.yml that scopes kubeapilinter to the api/ directory only, so it doesn't flag ordinary internal Go structs under pkg/. The rules with a large number of pre-existing violations (commentstart, optionalorrequired, optionalfields, arrayofstruct, ssatags) are disabled for now and can be enabled incrementally in follow-ups. I did fix the smaller, more meaningful set right away: the statusoptional and conditions rules. That meant marking the TestRunStatus and PrivateLoadZoneStatus fields as optional and adding the standard listType/patchStrategy markers to the Conditions fields, and I regenerated the CRDs accordingly.

make lint now passes cleanly. Let me know if you'd prefer a different scope for the first step, or if you'd rather see more rules enabled in this PR before merging. Happy to adjust.

moko-poi committed:
ci: address Bugbot feedback on helm CRDs and linter rebuild

- Sync the status.conditions listType/listMapKeys markers into the Helm
  chart CRD templates so Helm installs expose the same OpenAPI schema as
  config/crd/bases (keeps server-side apply/merge behavior consistent
  across install paths).
- Rebuild the custom golangci-lint binary whenever .custom-gcl.yml changes
  instead of only when the binary is missing, so plugin/version bumps are
  not silently ignored by a stale binary.

moko-poi commented on Makefile:
```diff
@@ -253,9 +253,13 @@ $(ENVTEST): $(LOCALBIN)
+		echo "Building golangci-lint with custom plugins..." ;\
+		GOBIN=$(LOCALBIN) go install github.com/golangci/golangci-lint/v2/cmd/golangci-lint@$(GOLANGCI_LINT_VERSION) ;\
+		$(LOCALBIN)/golangci-lint custom ;\
+	fi
```
Good catch. Fixed in 596e0e2: the target now depends on .custom-gcl.yml and drops the `if [ ! -f ]` guard, so the binary is rebuilt whenever the plugin config or version changes instead of being skipped when a stale binary is present.

moko-poi commented on config/crd/bases/k6.io_testruns.yaml:
```diff
@@ -6203,6 +6203,9 @@ spec:
                 type: array
+                x-kubernetes-list-map-keys:
+                - type
+                x-kubernetes-list-type: map
```
Fixed in 596e0e2. I synced the same listType=map / listMapKeys=[type] markers for status.conditions into the Helm chart CRD templates (charts/k6-operator/templates/crds/testrun.yaml and plz.yaml), so Helm installs now expose the same schema as config/crd/bases.

yorugac commented on .custom-gcl.yml:
```diff
@@ -0,0 +1,6 @@
+destination: ./bin
+plugins:
+- module: 'sigs.k8s.io/kube-api-linter'
+  version: 'v0.0.0-20251112164541-d94382a24f06'
```
This seems pretty old?
Looking at: https://pkg.go.dev/sigs.k8s.io/kube-api-linter?tab=versions

moko-poi replied:
@yorugac Good point. I bumped it to the latest available pseudo-version (2026-05-18)

yorugac commented on Makefile:
```diff
@@ -253,9 +253,13 @@ $(ENVTEST): $(LOCALBIN)
+golangci-lint: $(GOLANGCI_LINT) ## Build golangci-lint with kube-api-linter plugin.
+# Rebuild whenever .custom-gcl.yml changes (e.g. plugin or version bumps) so a
+# stale binary is never reused with an outdated plugin set.
+$(GOLANGCI_LINT): $(LOCALBIN) .custom-gcl.yml
```
Can this be written re-using `go-install-tool`?

moko-poi replied:
@yorugac Done. It turned out the lint job was failing because `golangci-lint custom` was writing over the running bootstrap binary (text file busy on Linux), so following the kube-api-linter docs (https://github.com/kubernetes-sigs/kube-api-linter?tab=readme-ov-file#golangci-lint-module) I now give the custom binary a distinct name (`golangci-lint-kube-api-linter`). That removes the collision and let me reuse `go-install-tool` for the bootstrap install as you suggested.

yorugac reviewed (changes requested):
@moko-poi, thank you. I kind of hoped that kube-api-linter would have some kind of release at least, but it seems like it's quite a bit more involved to add it 😅 I agree with the decisions on cutting scope that you made :+1: 
 

I have 2 request changes: please see the comments. Also, the lint step in CI doesn't work yet:

<img width="1007" height="216" alt="Image" src="https://github.com/user-attachments/assets/19a8e837-a964-4ef3-8113-47a32c7038ba" />

From: https://github.com/grafana/k6-operator/actions/runs/28075683570/job/84513210187 

I didn't yet try to run this locally, but is it trying to overwrite the currently executing binary? At least that's how the error looks like.

moko-poi committed:
ci: upgrade kube-api-linter to v0.0.0-20260518104151

Bump the kube-api-linter plugin from the 2025-11-12 snapshot to the
latest 2026-05-18 pseudo-version (kube-api-linter still has no tagged
release). The newer version adds the nonpointerstructs and defaults
rules, which report pre-existing violations in the v1alpha1 types, so
they are added to the staged-rollout disable list to be fixed in
follow-ups.

moko-poi committed:
fix(ci): avoid overwriting the running golangci-lint binary

The lint job failed on Linux with:

  build process: move golangci-lint binary: create destination file:
  open bin/golangci-lint: text file busy

`golangci-lint custom` wrote its output to bin/golangci-lint, the same
path as the bootstrap binary that was executing the `custom` command, so
overwriting the running executable failed with ETXTBSY. (It did not
reproduce on macOS, which allows overwriting a running binary.)

Follow the kube-api-linter docs and name the custom binary distinctly
(golangci-lint-kube-api-linter) so it never collides with the bootstrap
golangci-lint. The bootstrap install now reuses the go-install-tool
helper for consistency with the other tools.

yorugac reviewed (approved):
@moko-poi, thank you for the updates and clarity of explanations! I think Makefile is cleaner now and it's easier to follow the logic 🙌
